### PR TITLE
Fix azure files removal

### LINKED ISSUE #18436: Azure repository does not remove temporary dir
**Elasticsearch version**: tested with 2.4.0-SNAPSHOT

**Description of the problem including expected versus actual behavior**: 

When you create your azure repository, you get a message `The specified blob does not exist`

**Steps to reproduce**:

 1. `bin/plugin install cloud-azure`
 2. modify `config/elasticsearch.yml` to set `cloud.azure.storage.my_account.account: xxx` and `cloud.azure.storage.my_account.key: yyy`
 3. `bin/elasticsearch`
 4. Run the following script:

```sh
# Clean existing repo if any
curl -XDELETE localhost:9200/_snapshot/my_backup1?pretty

# Create azure repo
# This call gives in log: The specified blob does not exist
curl -XPUT localhost:9200/_snapshot/my_backup1?pretty -d '{
  "type": "azure"
}'
```

**Provide logs (if relevant)**:

```
[2016-05-18 11:03:24,914][WARN ][org.elasticsearch.cloud.azure.blobstore] [azure] can not remove [tests-ilmRPJ8URU-sh18yj38O6g/] in container {elasticsearch-snapshots}: The specified blob does not exist.
```

Note that there is a trailing `/` at the end which might explain the issue.

### PR DESCRIPTION
**Pull request by dadoonet**
Probably when we updated Azure SDK, we introduced a regression.
Actually, we are not able to remove files anymore.

For example, if you register a new azure repository, the snapshot service tries to create a temp file and then remove it.
Removing does not work and you can see in logs:

```
[2016-05-18 11:03:24,914][WARN ][org.elasticsearch.cloud.azure.blobstore] [azure] can not remove [tests-ilmRPJ8URU-sh18yj38O6g/] in container {elasticsearch-snapshots}: The specified blob does not exist.
```

This fix deals with that. It now list all the files in a flatten mode, remove in the full URL the server and the container name.

As an example, when you are removing a blob which full name is `https://dpi24329.blob.core.windows.net/elasticsearch-snapshots/bar/test` you need to actually call Azure SDK with `bar/test` as the path, `elasticsearch-snapshots` is the container.

To run the test, you need to pass some parameters: `-Dtests.thirdparty=true -Dtests.config=/path/to/elasticsearch.yml`

Where `elasticsearch.yml` contains something like:

```
cloud.azure.storage.default.account: account
cloud.azure.storage.default.key: key
```

Related to #16472
Closes #18436.

This PR also fixes a missing default value for setting `repositories.azure.container` and adds more logs.

### CODE REVIEW

**Comment by dadoonet** [comment]
@imotov Could you give a look at this?

File: plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/blobstore/AzureBlobContainer.java [hunk 1]
```diff
@@ -62,6 +62,7 @@ public AzureBlobContainer(String repositoryName, BlobPath path, AzureBlobStore b
 
     @Override
     public boolean blobExists(String blobName) {
+        logger.debug("blobExists({})", blobName);
```
Comment (imotov) [inline plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/blobstore/AzureBlobContainer.java]: I think logging these operations on the trace level would be more appropriate.

File: plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageServiceImpl.java [hunk 2]
```diff
@@ -182,13 +182,20 @@ public void deleteFiles(String account, LocationMode mode, String container, Str
         CloudBlobClient client = this.getSelectedClient(account, mode);
         CloudBlobContainer blob_container = client.getContainerReference(container);
```
Comment (imotov) [inline plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageServiceImpl.java]: Shouldn't this be `blobContainer`?

File: plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageServiceImpl.java [hunk 3]
```diff
@@ -182,13 +182,20 @@ public void deleteFiles(String account, LocationMode mode, String container, Str
...
+                logger.trace("removing blob [{}] full URI was [{}]", blobName, blobItem.getUri());
+                deleteBlob(account, mode, container, blobName);
             }
         }
     }
 
+    public static String blobNameFromUri(String container, URI uri) {
+        String path = uri.getPath();
+        // We remove the container name from the path
+        return path.replaceFirst("/" + container + "/", "");
```
Comment (imotov) [inline plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageServiceImpl.java]: For some reason this feels hacky... Especially, passing the container name. Maybe just chop at the second "/"?

**Comment by imotov** [comment]
Left a few minor comments. Otherwise LGTM.

**Comment by dadoonet** [comment]
@imotov I pushed some other changes. Do you mind giving a final review for it?

Thanks!

**Comment by dadoonet** [comment]
@clintongormley I believe I should backport this to 2.x branch right?
And probably to 1.7 as well as I believe latest 1.7 versions are also broken.

**Comment by clintongormley** [comment]
> @clintongormley I believe I should backport this to 2.x branch right? And probably to 1.7 as well as I believe latest 1.7 versions are also broken.

Yes please

File: plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageServiceImpl.java [hunk 4]
```diff
@@ -182,13 +182,34 @@ public void deleteFiles(String account, LocationMode mode, String container, Str
...
+     * Extract the blob name from a URI like https://myservice.azure.net/container/path/to/myfile
+     * It should remove the container part (first part of the path) and gives path/to/myfile
+     * @param uri URI to parse
+     * @return The blob name relative to the container
+     */
+    public static String blobNameFromUri(URI uri) {
+        String path = uri.getPath();
+
+        // We remove the container name from the path
+        // The 3 magic number cames from the fact we have // in the first part of the URI (protocol)
```
Comment (imotov) [inline plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageServiceImpl.java]: I don't get this comment. If you are asking for "path", shouldn't it return only /container/path/to/myfile with no "//" in the protocol?

**Comment by imotov** [comment]
Left a minor comment about a comment. I still think `blob_container` doesn't match our naming convention and should be renamed to `blobContainer`.

**Comment by dadoonet** [comment]
> Left a minor comment about a comment. I still think blob_container doesn't match our naming convention and should be renamed to blobContainer.

I do agree. I just missed the comment! Sorry!

File: plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageServiceImpl.java (hunk 4 above)
Comment (dadoonet) [inline plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageServiceImpl.java]: You are totally right. I wrote the comment before I implemented and added test then I forgot to update the comment.

Comment (dadoonet) [inline plugins/repository-azure/src/main/java/org/elasticsearch/cloud/azure/storage/AzureStorageServiceImpl.java]: Totally right! I'll fix.

**Comment by dadoonet** [comment]
For the record:

* 2.x: https://github.com/elastic/elasticsearch/issues/18571
* 1.7: https://github.com/elastic/elasticsearch-cloud-azure/issues/117

**Pull request merged**